BUILD

```diff
@@ -258,22 +258,60 @@ cc_binary(
 ################################################################################
 genrule(
     name = "generate_java_descriptor_proto",
-    tools = [":protoc"],
-    srcs = [ "src/google/protobuf/descriptor.proto", ],
-    outs = [ "com/google/protobuf/DescriptorProtos.java" ],
+    srcs = ["src/google/protobuf/descriptor.proto"],
+    outs = ["com/google/protobuf/DescriptorProtos.java"],
     cmd = "$(location :protoc) --java_out=$(@D)/../../.. $<",
+    tools = [":protoc"],
 )
 
 java_library(
     name = "java_proto",
-    visibility = ["//visibility:public"],
     srcs = glob([
-        "java/src/main/java/com/google/protobuf/*.java"
+        "java/src/main/java/com/google/protobuf/*.java",
     ]) + [
-      ":generate_java_descriptor_proto",
-    ]
+        ":generate_java_descriptor_proto",
+    ],
+    visibility = ["//visibility:public"],
+)
+
+################################################################################
+# Python support
+################################################################################
+
+# A hack that moves Python runtime srcs files form python/google/protobuf/... to
+# bazel-genfiles/google/protobuf/... so that the generated files can correctly
+# import modules from google.protobuf.
+PYTHON_SRCS = glob(
+    ["python/google/protobuf/**/*.py"],
+    exclude = [
+        "python/google/protobuf/**/*_pb2.py",
+        "python/google/protobuf/internal/*_test.py",
+        "python/google/protobuf/internal/test_util.py",
+    ],
+)
+
+PYTHON_MOVED_SRCS = [x[len("python/"):] for x in PYTHON_SRCS]
+
+PYTHON_RUNTIME_SRCS = PYTHON_MOVED_SRCS + \
+                      [x[:-len(".proto")] + "_pb2.py" for x in WELL_KNOWN_PROTOS]
+
+genrule(
+    name = "generate_python_protos",
+    srcs = PYTHON_SRCS + ["src/" + x for x in WELL_KNOWN_PROTOS],
```

Comment:
Hmm, I see. Is it not possible to have a separate BUILD file for the Python directory, so that pathnames are relative to python/?

If not, I could see why this hack is necessary.